### ISSUE TITLE
condition length& wav length

### ISSUE THREAD

**Issue by HallidayReadyOne** [bfs18/nsynth_wavenet#13]
Hi @bfs18 , in order to ensure that the condition length is consistent with the audio length, the code here 
https://github.com/bfs18/nsynth_wavenet/blob/1c276ea9f8f98d25aad7d9049c6e03101429ac53/wavenet/wavenet.py#L76
should be: 
cond = tf.slice(cond, [0, left_tl, 0], [-1, **cond_len - left_tl**, -1])  ?

**Comment by bfs18** [comment]
First, the third argument of [tf.slice](https://www.tensorflow.org/api_docs/python/tf/slice) is not `end position` but `size`.
Second, because of some implementation details, the predicted wave is a little longer than the original wave.
https://github.com/bfs18/nsynth_wavenet/blob/1c276ea9f8f98d25aad7d9049c6e03101429ac53/wavenet/masked.py#L186

**Comment by HallidayReadyOne** [comment]
"First, the third argument of tf.slice is not end position but size."
Yes, my mistake.

“Second, because of some implementation details, the predicted wave is a little longer than the original wave.” 
Assume that the original wave length is 9.  After shift_right(ori_wave), pad a zero on the left and remove the last sample. The audio length is still 9. 
Using filter_width = 3,  after https://github.com/bfs18/nsynth_wavenet/blob/1c276ea9f8f98d25aad7d9049c6e03101429ac53/wavenet/wavenet.py#L208
the output is still 9.
The conv1d in masked.py uses pad to keep the output length == input lenght,
https://github.com/bfs18/nsynth_wavenet/blob/1c276ea9f8f98d25aad7d9049c6e03101429ac53/wavenet/masked.py#L202
So I think the predicted wave has the some length as original wave. And because of this, we can use cate_targets (labels) and out (logits)  to compute the ce_loss directly.
What do you think?

**Comment by bfs18** [comment]
I'm not sure I understand your question.
At training, the wave length is given, upsampled mel condition data can be cropped to that length. At prediction, the wave length is unknown, we can only calculated the wave length from the mel condition data.

**Comment by HallidayReadyOne** [comment]
Oh, I misunderstood what you mean. The  “predicted wave”  you mentioned is the output of testing stage (not the output of the training phase). So during training, the input length is 7680, and the output length is 7680 too?  
https://github.com/bfs18/nsynth_wavenet/blob/1c276ea9f8f98d25aad7d9049c6e03101429ac53/wavenet/wavenet.py#L272

However I run the test_wavenet.py,  the input length is 7680, but the lenght of rt&ct is 11405, why not 7680?
print(len(set(rt.flatten().tolist())))  ->11405
print(len(set(ct.flatten().tolist())))  ->11405


Thank you for you kindly reply.

**Comment by HallidayReadyOne** [comment]
Oh, I misunderstood what you mean. The “predicted wave” you mentioned is the output of testing stage (not the output of the training phase). So during training, the input length is 7680, and the output length is 7680 too?

**Comment by HallidayReadyOne** [comment]
Oh, I misunderstood what you mean. The “predicted wave” you mentioned is the output of testing stage (not the output of the training phase). So during training, the input length is 7680, and the output length is 7680 too? 

And why choose 7680 for the input length, just because it is a multiple of 512?

**Comment by bfs18** [comment]
You misunderstand those lines. 
```
print(len(set(rt.flatten().tolist())))
print(len(set(ct.flatten().tolist())))
```
prints the number of distinct values of wave sample points. For example, If mu law is used, there are 2 ** 8 distinct values at most. Because round and floor operations may cause  that the number of cate target  is different from that of real target. So I printed the number. It has nothing to do with data shape.

-- So during training, the input length is 7680, and the output length is 7680 too?
Yes, you can see the tensor shape at runtime.
![screenshot from 2018-07-30 19-35-37](https://user-images.githubusercontent.com/39617376/43395253-018e6732-9430-11e8-930a-cb5e0623492d.png)

**Comment by HallidayReadyOne** [comment]
Thank you! 
I tried to output the tensor shape and deleted the previous comment (It doesn't seem to be completely removed, Orz).
Now I think I have figured it out. Thank you again for your answer！
